v2/cmd/controller/app/setup.go

```diff
@@ -0,0 +1,352 @@
+func SetupControllerManager(setupLog logr.Logger, flgs flags, ctx context.Context) manager.Manager {
+	scheme := controllers.CreateScheme()
+	_ = apiextensions.AddToScheme(scheme) // Used for managing CRDs
+
+	cfg, err := config.ReadAndValidate()
+	if err != nil {
+		setupLog.Error(err, "unable to get env configuration values")
+		os.Exit(1)
+	}
+
+	var cacheFunc cache.NewCacheFunc
+	if cfg.TargetNamespaces != nil && cfg.OperatorMode.IncludesWatchers() {
+		cacheFunc = cache.MultiNamespacedCacheBuilder(cfg.TargetNamespaces)
+	}
+
+	k8sConfig := ctrl.GetConfigOrDie()
+	mgr, err := ctrl.NewManager(k8sConfig, ctrl.Options{
+		Scheme:                 scheme,
+		MetricsBindAddress:     flgs.metricsAddr,
+		NewCache:               cacheFunc,
+		LeaderElection:         flgs.enableLeaderElection,
+		LeaderElectionID:       "controllers-leader-election-azinfra-generated",
+		Port:                   9443,
+		HealthProbeBindAddress: flgs.healthAddr,
+	})
+
+	if err != nil {
+		setupLog.Error(err, "unable to create manager")
+		os.Exit(1)
+	}
+
+	clients, err := initializeClients(cfg, mgr)
+	if err != nil {
+		setupLog.Error(err, "failed to initialize clients")
+		os.Exit(1)
+	}
+
+	crdManager, err := newCRDManager(clients.log, mgr.GetConfig(), cfg.PodNamespace)
+	if err != nil {
+		setupLog.Error(err, "failed to initialize CRD client")
+		os.Exit(1)
+	}
+
+	nonReadyResources, err := crdmanagement.GetNonReadyCRDs(ctx, cfg, crdManager)
+	if err != nil {
+		setupLog.Error(err, "failed to find non-ready CRDs")
+		os.Exit(1)
+	}
+
+	if len(flgs.crdPatterns) > 0 {
+		err = crdManager.ApplyCRDs(ctx, crdmanagement.CRDLocation, cfg.PodNamespace, flgs.crdPatterns)
+		if err != nil {
+			setupLog.Error(err, "failed to apply CRDs")
+			os.Exit(1)
+		}
+	}
+
+	if cfg.OperatorMode.IncludesWatchers() {
+		err = initializeWatchers(nonReadyResources, cfg, mgr, clients)
+		if err != nil {
+			setupLog.Error(err, "failed to initialize watchers")
+			os.Exit(1)
+		}
+	}
+
+	if cfg.OperatorMode.IncludesWebhooks() {
+		objs := controllers.GetKnownTypes()
+
+		objs, err = crdmanagement.FilterKnownTypesByReadyCRDs(clients.log, scheme, nonReadyResources, objs)
+		if err != nil {
+			setupLog.Error(err, "failed to filter known types by ready CRDs")
+			os.Exit(1)
+		}
+
+		if errs := generic.RegisterWebhooks(mgr, objs); errs != nil {
+			setupLog.Error(err, "failed to register webhook for gvks")
+			os.Exit(1)
+		}
+	}
+
+	// Healthz liveness probe endpoint
+	err = mgr.AddHealthzCheck("healthz", healthz.Ping)
+	if err != nil {
+		setupLog.Error(err, "Failed setting up healthz check")
+		os.Exit(1)
+	}
+
+	// Readyz probe endpoint
+	err = mgr.AddReadyzCheck("readyz", healthz.Ping)
+	if err != nil {
+		setupLog.Error(err, "Failed setting up readyz check")
+		os.Exit(1)
+	}
+
+	return mgr
+}
+
+type flags struct {
+	metricsAddr          string
+	healthAddr           string
+	enableLeaderElection bool
+	crdPatterns          []string
+}
+
+func ParseFlags(args []string) (flags, error) {
```

Comment:
`ParseFlags` is exported here which means that `flags` type should be too.

Reply:
Considering moving `Flags` into a separate file.